### PR TITLE
lib/model, lib/protocol: Fix error handling when terminating a connection

### PR DESCRIPTION
**Pull request by imsodin**
### Purpose

In the protocol's readerloop do actually set the err variable instead of
returning directly, such that it is correctly passed to c.close in the deferred
statement.
In the model do a nil-check before calling err.Error().

### Testing

Simple unit test that doesn't panic anymore.

### CODE REVIEW

File: lib/protocol/protocol.go [hunk 1]
```diff
@@ -340,11 +340,12 @@ func (c *rawConnection) readerLoop() (err error) {
 	for {
 		select {
 		case <-c.closed:
-			return ErrClosed
```
Comment (calmh) [inline lib/protocol/protocol.go]: As far as I understand, these changes should be an ugly no-op?

File: lib/model/model.go [hunk 2]
```diff
@@ -1270,10 +1270,15 @@ func (m *Model) Closed(conn protocol.Connection, err error) {
 	m.pmut.Unlock()
 
 	l.Infof("Connection to %s at %s closed: %v", device, conn.Name(), err)
-	events.Default.Log(events.DeviceDisconnected, map[string]string{
-		"id":    device.String(),
-		"error": err.Error(),
-	})
+
+	data := map[string]string{
```
Comment (calmh) [inline lib/model/model.go]: It should be safe to assume that we never close with a nil error - if the error was nil, the connection would not have been closed. I'd even throw in an `if err == nil { panic("bug:...") }` to that effect. Clearly you've *seen things* I haven't, though?

**Comment by imsodin** [comment]
@calmh Yes, I have seen ghosts aka this nil deref https://github.com/syncthing/syncthing/issues/5043#issuecomment-411097106 which points to https://github.com/syncthing/syncthing/blob/master/lib/model/model.go#L1275. The return value `err` is declared in the function signature and used it in the defer statement, so I assume if we return directly that error doesn't get assigned to `err` and thus the defer statement uses a nil error. On second thoughts, that can't be because then every connection close should have resulted in a nil deref.

So actually I am at a loss as to how the referenced nil deref every happened...

File: lib/model/model.go (hunk 2 above)
Comment (AudriusButkevicius) [reply lib/model/model.go]: Agree. The bug is clearly in the protocol package.

**Comment by imsodin** [comment]
I now believe my changes in this PR are BS from A to Z. In the original report `model.Closed` was called with a nil error, but I don't understand at all how. It also looks like that was a double panic, something really strange is going on.
```
github.com/syncthing/syncthing/lib/model.(*Model).Closed(0xc420211320, 0x7f3481835e60, 0xc4224d3710, 0x0, 0x0)
```

Also some further weirdness, that probably explains the discrepancy between my example above (direct defer) vs the code in protocol.go (defer with func):

```golang
package main

import (
	"fmt"
)

func deferStuff() (err error) {
	defer func() {
		fmt.Println("defer func", err)
	}()
	defer fmt.Println("defer direct", err)
	return fmt.Errorf("this is an error")
}

func main() {
	fmt.Println(deferStuff())
}
```

```
defer direct <nil>
defer this is an error
this is an error
```

**Comment by calmh** [comment]
If we panicked in the protocol code while `err` was still `nil`, the defer would be called with `c.close(nil)`. I think this is an unexpected side effect of the panic, but probably doesn't matter much as we're already going down due to the panic... (which should be fixed, of course)

**Pull request closed without merging**